### ISSUE TITLE
Pyright's detection of calls to abstract methods seems to be unsound

### ISSUE THREAD

**Issue by mheiber** [microsoft/pyright#8550]
**Describe the bug**

Pyre's detection of calls to abstract methods seems to be unsound, see code example below

**Code or Screenshots**

```python
from abc import ABC, abstractmethod

class A(ABC):
    @classmethod
    @abstractmethod
    def abs(cls) -> None:
        raise NotImplementedError()

    @classmethod
    def foo(cls) -> None:
        cls.abs()
        print(cls)


class B(A):
    pass


A.abs() # pyright error
# But pyright doesn't doesn't catch the same problem when we add some indirection:
B.foo() # raises NotImplementedError() at runtime. pyright doesn't catch this
```

**VS Code extension or command-line**

pyright-playground.net as of 2024-07-26: [pyright playground link](https://pyright-play.net/?code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEAaWOAZyxBgSwgFMsALMAEwCh2EAbGc8vAAoChAJQAudlGlQAAjz7kGzNlJmz4larWUsOMqK3rAy5QT3KioAWgB8UAHJgAdvUkGD1ZOXqOwWAEl0bnoGZyx6VgBREHAQQVFOA3lefl1VAyMTYDAwc25LG3snV3cPaQsAOk0EtXK0EGRw-MtOLlSBfGEJOqg0RTb2XGqKBKgAYj6AT0aAcyYceljMdkn8AFccNBnkeZxWMHpyZwByfcPjs6gEGCwEJihmX3IYBj7wOBDoAHcmemcoN9fDBWKwoORIL4mqxkCB6LRkC5JPhKjk8lZJl4fAInIFgqF-hFost4lZblAQOtwqh6JVpnMFoYLqccDc7g9mN4gA). Sorry, I could not find where the version is specified, in the playground nor in the github repo for the playground.

**Comment by erictraut** [comment]
Missing abstract methods are detected at the time you attempt to instantiate a class, call a class method or static method that is marked abstract, or if you mark a class as `@final` and it doesn't implement all of the abstract methods of its parent.

I don't see a way to detect the bug in your code without performing global (cross-function) type analysis, which no type checkers do (for good reason). Even with global analysis, this would be difficult to detect in many cases. Not surprisingly, mypy does not report any error for `B.foo()` either.

My recommendation is to get into the habit of marking your leaf classes as `@final` if you want a static type checker to help you detect problems like this one.

**Comment by mheiber** [comment]
Thanks for the suggestion re @final, and I understand there's no easy win solution. Here are a few options to make this sound if pyright project revisits in future (which don't require interprocedural analysis):
- Don't allow calls to abstract static methods except when `cls` is lexically in a non-abstract class. This is part of an approach we're considering for Hack, to plug a similar soundness hole.
- Don't allow methods to be both static and abstract: this is what TypeScript does.

**Comment by erictraut** [comment]
Thanks for the suggestions. 

I'm not sure I understand the first suggestion. Do you mean "lexically in a non-abstract *method*"?

The second suggestion would need to be extended to class methods as well as static methods. And even then, there'd still be a hole with instance methods that are called directly from a class (supplying a `self` value manually).

Unfortunately, I think that both of your suggestions would result in quite a few false positives in real-world Python code. I also don't think this is a major source of bugs. For those reasons, it's probably not a good tradeoff for pyright users.

If you think I'm underestimating the frequency of bugs of this nature or overestimating the false positives such a change would cause and can provide some evidence (data or anecdotes) that backs that up, I'm open to exploring these options further.

**Comment by mheiber** [comment]
I agree any solution would likely have a lot of cases where the type-checker says there's a problem but the bad thing (calling an abstract method) doesn't happen.

I don't have any data for Python and fwiw completely agree that it's an empirical question how bad the disease is and how expensive the cure would be.

Re the first suggestion: sorry, what I wrote before was wrong. What would be needed for soundness is to show that, in all allowed cases of `cls.foo`, `cls` is concrete. 

Here's another try at a solution sketch, admittedly burdensome:
- disallow `C.foo()` where `C` is the name of an abstract class and `foo` is a class method
    - iuc in `D = A` where `A` is a class, D is of type `Type[A]`. If we wanted full safety, then`Type` would probably have to be bifurcated into `ConcreteType[T]` and `AbstractType[T]` such that calls to class methods of `T` are only allowed through `ConcreteType[T]`. For example, in `D.foo()` where `foo` is a class method, we would ensure that `D` is of type `ConcreteType[T]` for some T.
- For calls to unbound class methods, the first argument must be a concrete class.

> I didn't follow the point about instance methods: once you're in an instance method, you know the class is concrete (assuming we prevent abstract classes from being instantiated).

**Comment by erictraut** [comment]
Here's [an illustration](https://pyright-play.net/?reportMissingModuleSource=false&code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEAaWOAZyxBgSwgFMsALMAEwCh2EAbGc8vAAoChAJQAudlGlQAAvErVaDZmykzW9YGXKDy9bsFFQAtAD4oAOTAA7epJmOo1ZPqtgsASXTd6DG1j0rACiIOAggqKcjprawGBgegZGphbWdg5O0vqGAHQKkZxcvPxQ%2BMIS6tJofORFPLVQhIL4lTFaOkmGxubuGVVONfycSAC8TYX4ufGJCFHsQA) of what I mean about instance methods.

**Comment by mheiber** [comment]
> No pressure to reply, only asking out of curiosity

Thanks @erictraut: I couldn't follow what the issue with instance methods is, though. Afaict there's no runtime error in the code sample you shared, and (given that pyright enforces that abstract classes can't be instantiated) the class will be concrete in the example you shared.

**Comment by mheiber** [comment]
I think Pyright's handling of abstract methods could be made sound by separating the `Type` static type into `ConcreteType` and `AbstractType`.

> This was just for curiosity, and I doubt is worth the cost of a breaking change

Here's a sketch, considering only static methods with 1 parameter.
This is simplified compared to full Python (of course), but I think covers the kind of case in the issue description, given that we can consider class methods sugar for static methods that take a parameter of type `Type[name_of_the_class]` ([example](https://pyright-play.net/?code=GYJw9gtgBAhgRgYygSwgBzCALlAggIQGEAaWOAZyxBgSwgFMsALMAEwChRIosBPNZADsA5inSYcAFX7127BABsY5cngAUBQgEoAXOygGoAAUowsyBA2Zt9h1vWBlyarVAC0APigA5MIPp6hkFQ1Mjk9D5gWACS6Ar0DIJY9KwAoiDgIC5yQSZYZhZWLBxB9o7AYGBqANZKKjpQ0mj0ANoA5LhtALqunpH%2BgcEGtcrkAHTwzlpy8nWq%2BBq6tgZoozPsuBMULlAAxFBovCDIwkw49BmY7Pv4AK44h8enOKxg9OSCbS9vH19QCGYEEweEwIuQYAwDuA4PFoAB3UGCKBwiIwVisKDkSARISsZAgei0ZB%2BPT4MYVKq4Vz7ULhVS%2BGJxBL0JIpdKZHZmEK3JKoehjA5HE5nKCvd6fHAArBAkFhIA)).

```
T inherits from U
----------------------------------
AbstractType[T] <: AbstractType[U]

T inherits from U
----------------------------------
ConcreteType[T] <: ConcreteType[U]

T inherits from U
----------------------------------
ConcreteType[T] <: AbstractType[U]

T inherits from U      U is a concrete class
----------------------------------------------
AbstractType[T] <: ConcreteType[U]


Var is a concrete class or Var.foo is a concrete method   foo: Callable[[], T]
--------------------------------------------------------------------------
Var.foo()  : T
```

Here's a model in Alloy with the correctness property that there are no calls to abstract methods: https://gist.github.com/mheiber/a42f24efe766ef532c29c3e5054361cb

**Comment by mheiber** [comment]
Here's an approach adapted from something I'm considering for Hack ([interactive spec](https://github.com/mheiber/lsbsteak/)). I don't know if it's practical adopt for Pyright, including here just in case it's of interest.



Declarative typing:

```
T <: U                                                  
----------------------------------------------------    
Type<T> <: Type<U>                            


T <: U                              
--------------------------------------------
ConcreteType<T> <: ConcreteType<U> 


------------------------------------------------------------------
ConcreteType<T> <: Type<T>



C is a concrete class
-------------------------------------   
C : ConcreteType<C>        


C is a class
-------------------------------------------------
C : Type<C>    
```

Code from the issue description, with pyright now catching the error:

```
from abc import ABC, abstractmethod

class A(ABC):
    @classmethod
    @abstractmethod
    def abs(cls) -> None:
        raise NotImplementedError()

    @classmethod
    def foo(cls) -> None:
        cls.abs() # Error, cls must be concrete in order to call abstract methods.
                  # Consider updating the param: `cls: ConcreteType[A]`


class B(A):
    pass


A.abs() # pyright error
B.foo() # raises NotImplementedError() at runtime if the pyright error above is ignored
```

The author of A can prevent some runtime errors like this (extended example):


```
from abc import ABC, abstractmethod

class A(ABC):
    @classmethod
    @abstractmethod
    def abs(cls) -> None:
        raise NotImplementedError()

    @classmethod
    def foo(cls: ConcreteType[A]) -> None:
        cls.abs() # OK


class B(A):
    pass

class C(B):
    # C is non-abstract, since all methods are implemented
    @classmethod
    def abs(cls) -> None:
        print(cls)



A.abs() # pyright error
B.foo() # pyright error
C.foo() # OK!
```
